File: backend/phase4/services/hermes_orchestrator/adapters/leaderboard_adapter.py
"""Leaderboard Service Adapter"""
from typing import Dict, Any, Optional
from .base import BaseAdapter

class LeaderboardAdapter(BaseAdapter):
    def __init__(self, base_url: Optional[str] = None):
        super().__init__("leaderboard", base_url)

    async def get_leaderboard(
        self,
        period: str = "weekly",
        language: Optional[str] = None,
        limit: int = 100
    ) -> Dict[str, Any]:
        params = {"period": period, "limit": limit}
        if language:
            params["language"] = language
        result = await self._request("GET", "/v1/leaderboard", params=params)
        return result.data if result.success else {"error": result.error}

    async def submit_score(self, user_id: str, score: int, category: str) -> Dict[str, Any]:
        result = await self._request("POST", "/v1/submit-score", payload={
            "user_id": user_id,
            "score": score,
            "category": category
        })
        return result.data if result.success else {"error": result.error}

File: mobile_app/lib/data/adapters/leaderboard_adapter.dart
import 'package:dio/dio.dart';
import 'package:hermes_linguamind/core/constants/api_endpoints.dart';
import 'package:hermes_linguamind/core/utils/logger.dart';
import 'package:hermes_linguamind/data/adapters/api_client.dart';
import 'package:hermes_linguamind/data/models/leaderboard_models.dart';

class LeaderboardAdapter {
  factory LeaderboardAdapter() => _instance;
  LeaderboardAdapter._();
  static final LeaderboardAdapter _instance = LeaderboardAdapter._();

  final ApiClient _client = ApiClient();

  Future<LeaderboardResponse> getLeaderboard({
    final String period = 'weekly',
    final int page = 1,
    final int pageSize = 20,
  }) async {
    try {
      final response = await _client.get<dynamic>(
        ApiEndpoints.leaderboard,
        queryParameters: {
          'period': period,
          'page': page,
          'page_size': pageSize,
        },
      );
      return LeaderboardResponse.fromJson(
        response.data as Map<String, dynamic>,
      );
    } on DioException catch (e) {
      AppLogger.error('Get leaderboard failed', e);
      throw _handleError(e);
    }
  }

  Future<void> submitScore({
    required final String userId,
    required final int xp,
    required final int streakDays,
  }) async {
    try {
      await _client.post<dynamic>(
        ApiEndpoints.submitScore,
        data: {'user_id': userId, 'xp': xp, 'streak_days': streakDays},
      );
    } on DioException catch (e) {
      AppLogger.error('Submit score failed', e);
      throw _handleError(e);
    }
  }

  Exception _handleError(final DioException e) {
    final response = e.response;
    if (response != null) {
      final data = response.data as Map<String, dynamic>?;
      final message =
          data?['detail'] as String? ??
          data?['message'] as String? ??
          'Leaderboard request failed';
      return Exception(message);
    }
    return Exception('Network error. Please check your connection.');
  }
}
